dace/frontend/fortran/ast_transforms.py

```diff
@@ -2365,7 +2358,22 @@ def visit_BinOp_Node(self, node: ast_internal_classes.BinOp_Node):
 
             node.type = lval_type
 
-        if node.op == '=' and isinstance(node.lval, ast_internal_classes.Name_Node) and node.rval.type != 'VOID' and node.lval.type == 'VOID':
+        is_assignment = isinstance(node.lval, (ast_internal_classes.Name_Node))
+
+        should_be_updated = False
+        if node.lval.type == 'VOID':
+            should_be_updated = True
+
+        if self._get_sizes(node.lval) is None:
+            should_be_updated = True
+
+        # We do NOT overwrite size of a varible if it has been defined by the user.
+        # We only do it for temporaries introduced by our transformations.
+        # At this moment, the only way to distinguish between them is the `tmp_` prefix.
+        if isinstance(node.lval, ast_internal_classes.Name_Node) and node.lval.name.startswith("tmp_"):
```

Comment:
Out of curiosity: In what scenario we would need to change the size of a temporary variable? The extracted variables' shapes are basically fully determined by the expression itself, aren't they?

Reply:
No, they aren't. Some of the transformations extract the expression when size is unknown, e.g., recursive calls to intrinsics. Another example is the processing of non-contiguous slices where the size of the result changes.

Reply:
So, in the first example: size can go from an unknown state to a known state, IIUC. That sounds reasonable. Does the second example mean that the size can go from a known state to a different known state? Could you perhaps give a fortran example of when that change is needed?

Reply:
The famous case of transposing is in `get_albedos`. The argument's size changes once we process the non-contiguous slicing.

Reply:
(Please feel free to submit; I'm happy as long as I get my answers)

So, say we have this expression `tmp = TRANSPOSE(this_var_378 % lw_emissivity(istartcol_var_380 : iendcol_var_381, config_var_379 % i_emiss_from_band_lw(config_var_379 % i_band_from_reordered_g_lw)))`. Do you mean that the size of `tmp` will depend on the **content** of the operands of `transpose()`, e.g., `config_var_379 % i_emiss_from_band_lw(config_var_379 % i_band_from_reordered_g_lw)`?

Reply:
It will depend on the shape of that array, not the contents. This, in turn, will recursively depend on the shape of `config_var_379 % i_band_from_reordered_g_lw`. We need to extract each component and process it sequentially to get the size of `tmp`.

Reply:
Great! So `tmp`'s shape is really in an unknown state until we know all the nested shapes, IIUC. Is there any case where we would want to change a known shape into a different want (e.g. `(1:5, 4:7)` to `(1:15)`, or `(b0:e0)` to `(b1:e1)` where `b0,e0,b1,e1` are symbols)?